File: 0033-search-in-rotated-sorted-array/0033-search-in-rotated-sorted-array.java
class Solution {
    public int search(int[] nums, int target) {
        int startIndex = 0;
        int endIndex = nums.length - 1;
        
        while (startIndex <= endIndex) {
            int middleIndex = (startIndex + endIndex) / 2;
            
            if (nums[middleIndex] == target) {
                return middleIndex;
            }
            
            if (nums[startIndex] <= nums[middleIndex]) {
                if (target < nums[startIndex] || target > nums[middleIndex]) {
                    startIndex = middleIndex + 1;
                } else {
                    endIndex = middleIndex - 1;
                }
            } else {
                if (target > nums[endIndex] || target < nums[middleIndex]) {
                    endIndex = middleIndex - 1;
                } else {
                    startIndex = middleIndex;
                }
            }
        }
        
        return -1;
    }
}

File: 0033-search-in-rotated-sorted-array/0033-search-in-rotated-sorted-array.js
/**
 * @param {number[]} nums
 * @param {number} target
 * @return {number}
 */
var search = function(nums, target) {
    let left = 0;
    let right = nums.length - 1;
    
    while (left <= right) {
        let middle = Math.floor((left + right) / 2);
        
        if (target === nums[middle]) {
            return middle;
        }
        
        if (nums[left] <= nums[middle]) {
            if (target > nums[middle] || target < nums[left]) {
                left = middle + 1;
            } else {
                right = middle;
            }
        } else {
            if (target < nums[middle] || target > nums[right]) {
                right = middle;
            } else {
                left = middle + 1;
            }
        }
    }
    
    return -1;
};